backend/s3: Document need to reconfigure backend when updating Terraform (#34098)
hashicorp/terraform: website/docs/language/upgrade-guides/index.mdx

The `run` block also applies or plans the main configuration by default, there is no need for the specific module call seen in the experimental framework.

## S3 Backend

The S3 Backend has been updated, with a number of added and deprecated fields.
These changes bring the S3 Backend configuration closer to the AWS Provider configuration.

-> **Note:** Because of these changes, you may need to run `terraform init -reconfigure`,
even if there have been no changes to your backend configuration.

### Deprecations

The major deprecations are discussed here.
For more information, consult the [S3 Backend documentation](/terraform/language/settings/backends/s3).

Configuration for assuming an IAM Role has been moved from a number of top-level attributes into the attribute `assume_role`.
Previously, the configuration to assume the IAM role `arn:aws:iam::123456789012:role/example` with a session name `example-session` and a session duration of 15 minutes was:

```hcl

> terraform {

Reviewer: should this example show backend S3 block? 
```hcl
terraform {
  backend "s3" {
    assume_role  {
        role_arn     = "arn:aws:iam::123456789012:role/example"
        session_name = "example-session"
        duration     = "15m"
    }
  }
}
```

Author: Yes, good catch

> terraform {
>   role_arn                     = "arn:aws:iam::123456789012:role/example"
>   session_name                 = "example-session"
>   assume_role_duration_seconds = 900
> }

Reviewer 2: ```suggestion
terraform {
  backend "s3" {
    # additional configuration omitted for brevity
    role_arn                     = "arn:aws:iam::123456789012:role/example"
    session_name                 = "example-session"
    assume_role_duration_seconds = 900
  }
}
```

```

The updated configuration is:

```hcl

> terraform {
>   assume_role {
>     role_arn     = "arn:aws:iam::123456789012:role/example"
>     session_name = "example-session"
>     duration     = "15m"
>   }
> }

Reviewer 2: ```suggestion
terraform {
  backend "s3" {
    # additional configuration omitted for brevity
    assume_role = {
      role_arn     = "arn:aws:iam::123456789012:role/example"
      session_name = "example-session"
      duration     = "15m"
    }
  }
}
```

-> **Note:** Because of these changes, you may need to run `terraform init -reconfigure`,
even if there have been no changes to your backend configuration.

### Deprecations

The major deprecations are discussed here.
For more information, consult the [S3 Backend documentation](/terraform/language/settings/backends/s3).

Configuration for assuming an IAM Role has been moved from a number of top-level attributes into the attribute `assume_role`.
Previously, the configuration to assume the IAM role `arn:aws:iam::123456789012:role/example` with a session name `example-session` and a session duration of 15 minutes was:

```hcl
terraform {
  backend "s3" {
    # additional configuration omitted for brevity
    role_arn                     = "arn:aws:iam::123456789012:role/example"
    session_name                 = "example-session"
    assume_role_duration_seconds = 900
  }
}
```

The updated configuration is:

```hcl
terraform {
  backend "s3" {
    # additional configuration omitted for brevity
    assume_role = {
      role_arn     = "arn:aws:iam::123456789012:role/example"
      session_name = "example-session"
      duration     = "15m"
    }
  }
}
```

> The AWS API endpoint override attributes
> `endpoint` (for S3),
> `dynamodb_endpoint`,
> `iam_endpoint`, and
> `sts_endpoint`
> have been replaced with the attributes `endpoints` and the corresponding nested attributes
> `s3`,
> `dynamodb`,
> `iam`, and
> `sts`.

Reviewer 3: ```suggestion
The following AWS API endpoint override attributes are now nested under the `endpoint` attribute:
- `s3`
- `dynamodb`
- `iam`
- `sts` 

The `endpoint` block replaces the following top-level attributes:
- `endpoint` (for S3),
- `dynamodb_endpoint`,
- `iam_endpoint`, and
- `sts_endpoint`
```
Trying to use lists to make this more scannable and easier to compare. We could also consider using a table, but I'm not sure it would work as a result of the change to the attribute hierarchy.
